### PR TITLE
Signature verification

### PR DESCRIPTION
**Pull request by tpepper**
This PR represents the final working state of Tudor and I as we went through OpenSSL's poor documentation and addressed review comments from PR #85 .  The src/signature.c:verify_signature() function is an abomination, but we think we doesn't leak openssl state and valgrind is agreeing.  The code is focused now on just the MoM signature verification flow, which means some functions are now static.  The cert and cert path are now autoconf configurable.

### CODE REVIEW

File: src/signature.c [hunk 1]
```diff
@@ -24,198 +24,375 @@
...
-}
-
-void signature_terminate(void)
-{
+	return result;
 }
 
-bool signature_verify(const char UNUSED_PARAM *data_filename, const char UNUSED_PARAM *sig_filename)
+#else
+bool download_and_verify_signature(const char *data_url, const char *data_filename)
```
Comment (phmccarty) [inline src/signature.c]: The UNUSED_PARAM macro can be used for the function args.

File: test/signature_verify_test.c [hunk 2]
```diff
@@ -46,19 +46,12 @@ int main(int argc, char **argv)
 		usage(argv[0]);
 	}
 
-	if (!signature_initialize(argv[3])) {
-		fprintf(stderr, "Can't initialize!\n");
-		exit(-1);
-	}
-
-	if (signature_verify(argv[1], argv[2])) {
+	if (verify_signature(argv[1], argv[2])) {
```
Comment (phmccarty) [inline test/signature_verify_test.c]: This test should be calling `download_and_verify_signature()`.

File: src/signature.c [hunk 3]
```diff
@@ -24,198 +24,375 @@
...
+	if (!validate_certificate()) {
+		goto fail;
+	}
 
-static bool initialized = false;
+	/* Push our trust cert(s) to the stack, which is a set of certificates
+	 * in which to search for the signer's cert. */
+	x509_stack = sk_X509_new_null();
+	if (!x509_stack) {
+		return false;
```
Comment (phmccarty) [inline src/signature.c]: This is a void function, so the above line should be updated.

Comment (tpepper) [inline src/signature.c]: fixed

File: src/signature.c (hunk 1 above)
Comment (tpepper) [inline src/signature.c]: fixed.

File: test/signature_verify_test.c (hunk 2 above)
Comment (tpepper) [inline test/signature_verify_test.c]: fixed

File: src/manifest.c [hunk 4]
```diff
@@ -494,7 +486,7 @@ static int try_delta_manifest_download(int current, int new, char *component, st
 /* TODO: This should deal with nested manifests better */
 static int retrieve_manifests(int current, int version, char *component, struct file *file)
 {
-	char *url;
```
Comment (tpepper) [inline src/manifest.c]: this appears unneeded

**Comment by phmccarty** [comment]
+1 from me

**Comment by bradTpeters** [comment]
Still reviewing but looking good thus far

File: src/manifest.c [hunk 5]
```diff
@@ -503,17 +495,19 @@ static int retrieve_manifests(int current, int version, char *component, struct
 
 	string_or_die(&filename, "%s/%i/Manifest.%s.tar", state_dir, version, component);
 	if (stat(filename, &sb) == 0) {
-		return 0;
+		ret = 0;
+		goto out;
```
Comment (phmccarty) [inline src/manifest.c]: I'm running into a segfault with this change. The "goto out" will lead to freeing "url", which is not initialized at this point.

File: configure.ac [hunk 6]
```diff
@@ -29,6 +25,19 @@ AS_IF([test "x$enable_bzip2" != "xno" ],
 	 AC_CHECK_LIB([bz2], [BZ2_bzBuffToBuffCompress], [], [AC_MSG_ERROR([the libbz2 library is missing])])],
   [AC_DEFINE(SWUPD_WITHOUT_BZIP2,1,[Do not use bzip2 compression])]
 )
+AC_ARG_ENABLE(
+  [signature-verification],
+  [AS_HELP_STRING([--enable-signature-verification], [Enable signature check (disabled by default)])],
+  [AC_DEFINE([SIGNATURES], [], [Enable signature check as default])]
```
Comment (bradTpeters) [inline configure.ac]: Is this enabled or disabled by default? The text here seems to be divided

File: src/signature.c [hunk 7]
```diff
@@ -24,198 +24,379 @@
...
+ *
+ * returns: true if can initialize and validate certificates, otherwise false */
+static void initialize_signature(void)
+{
+	ERR_load_crypto_strings();
+	ERR_load_PKCS7_strings();
+	if (!get_pubkey()) {
+		goto fail;
+	}
+	if (!validate_certificate()) {
```
Comment (bradTpeters) [inline src/signature.c]: This call should not be made - it's a no-op with the current self-signed cert

File: src/signature.c [hunk 8]
```diff
@@ -24,198 +24,379 @@
...
-		return NULL;
+	/* TODO: CRL and Chains are not required for the current setup, but we may
+	 * implement them in the future 
+	if (!crl) {
+		printf("No certificate revocation list provided\n");
 	}
-	if (X509_STORE_load_locations(store, ca_filename, ca_dirname) != 1) {
-		goto err;
+	if (!chain) {
+		printf("No certificate chain provided\n");
```
Comment (bradTpeters) [inline src/signature.c]: This seems fair, but we should describe the result of this, not simply the fact that we don't have a cert chain.

Comment (bradTpeters) [inline src/signature.c]: Could reword this to: "Proceeding with signature validate using self-signed certificate"

Comment (phmccarty) [inline src/signature.c]: This code is commented out right now...

File: configure.ac (hunk 6 above)
Comment (phmccarty) [inline configure.ac]: disabled by default. Passing the configure option as --enable-signature-verification enables it.

File: src/signature.c [hunk 9]
```diff
@@ -17,205 +17,426 @@
...
+	PKCS7 *p7;
+	BIO *verify_BIO;
+
+	if (!initialize_signature()) {
+		terminate_signature();
+		return false;
+	}
+
+	//this function has issues: EVP_add_digest(EVP_sha256());
+	OpenSSL_add_all_digests();
```
Comment (bradTpeters) [inline src/signature.c]: Here's a more limited digest add which seems to do the trick:

-       //this function has issues: EVP_add_digest(EVP_sha256());
-       OpenSSL_add_all_digests();
+       EVP_add_digest(EVP_sha1());
+       EVP_add_digest_alias(SN_sha1, "ssl3-sha1");
+       EVP_add_digest_alias(SN_sha1WithRSAEncryption, SN_sha1WithRSA);

Comment (tpepper) [inline src/signature.c]: Does that work on the second run through the function?

Our big issue this afternoon was appearing to be that we must only call EVP_add_digest() once at process start (not at each run through our verifier function) and must only call the clean up functions one on process termination (not at the end of each run through the verifier function).

**Comment by tpepper** [comment]
I moved the terminate_signature() and initialize_signature() into a new swupd_deinit() and the existing swupd_init().  I'm now able to verify, verify/fix, bundle add and bundle remove.

The pre-existing load_mom() segfault in swupd search is still there...didn't have time to look into it.  Am hoping @bradTpeters and @phmccarty made progress on it at the end of the day separately?

And I observe that bundle add/remove both have a bunch of pre-existing memory leakage...need a issue opened, run through valgrind and some thought on how to add proper memory management?

File: src/manifest.c (hunk 5 above)
Comment (phmccarty) [inline src/manifest.c]: "url" now initialized, so this issue is fixed.

**Comment by phmccarty** [comment]
There is a minor build issue when building without signature verification enabled, so I'll push a fix for that as a followup.

+1

**Comment by phmccarty** [comment]
@tpepper I'm not seeing the swupd search bug anymore. Let me know if you still see it.
